### ISSUE TITLE
[OptionsResolver] Options not provided by user triggers deprecation

### ISSUE THREAD

**Issue by ro0NL** [symfony/symfony#28848]
**Description**  
Since 4.2 OptionsResolver can deprecate options, however the deprecation is triggered from `offsetGet()`

https://github.com/symfony/symfony/blob/9bec1fcf08cecabb55681badaf6bfc728d1e4e20/src/Symfony/Component/OptionsResolver/OptionsResolver.php#L943

Im not sure about this, it causes the deprecation to trigger without even being provided by the user (i.e. when accessing the default value). Shouldnt it be moved to `resolve()` instead?

ref #27277 cc @yceruto

**Comment by ro0NL** [comment]
i confused `array $options` and `Options $options`. Of course only in case of the latter the deprecation  triggers when accessing the option, but this can still happen e.g. in SF core (e.g. from when referenced from a default value closure in another option)

So no real issue so far, but maybe it needs special API like `offsetGet($option, $triggerDeprecation = false)` sooner or later.

**Comment by yceruto** [comment]
> [OptionsResolver] Options not provided by user triggers deprecation

I agree, but this behavior is expected also for ppls who use a deprecated option during a lazy evaluation or normalization of another option. In those cases, they should always notice the deprecation message (even, if no value is provided).

Tomorrow I'll take a look,
Cheers!

**Comment by ro0NL** [comment]
See #28860 that includes a patch

IMHO `setDefaults` belongs to the definition side (along with `setDeprecated`), and only consuming deprecated options should trigger

**Comment by yceruto** [comment]
See the proposal to solve it https://github.com/symfony/symfony/pull/28878

**Comment by stof** [comment]
@ro0NL if you have a lazy default value using the value of another deprecated option, you need to be warned, to know you need to migrate your callable (but there should be a way to opt-out of the deprecation to be able to write a BC layer)

**Comment by ro0NL** [comment]
that will happen no? As soons as you access it thru `offsetGet()`?
